table/fib-strategy-hashtable.go

```diff
@@ -0,0 +1,410 @@
+/* YaNFD - Yet another NDN Forwarding Daemon
+ *
+ * Copyright (C) 2022 Danning Yu.
+ *
+ * This file is licensed under the terms of the MIT License, as found in LICENSE.md.
+ */
+
+// This file implements a hash table version of the FIB, as outlined in section 2.3 of:
+// W. So, A. Narayanan and D. Oran,
+// "Named data networking on a router: Fast and DoS-resistant forwarding with hash tables,"
+// Architectures for Networking and Communications Systems, 2013, pp. 215-225,
+// doi: 10.1109/ANCS.2013.6665203.
+
+package table
+
+import (
+	"sync"
+
+	"github.com/cespare/xxhash"
+	"github.com/named-data/YaNFD/ndn"
+)
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+
+	return b
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+type virtualDetails struct {
+	// isVirtual marks whether this a virtual node, as defined in the paper.
+	// If there exists a name with length greater than M, then the node is
+	// marked as virtual.
+	isVirtual bool
+
+	// md is the max depth associated with this virtual node, as defined in the paper.
+	md int
+}
+
+// FibStrategy Tree represents a tree implementation of the FIB-Strategy table.
+type FibStrategyHashTable struct {
+	// m is the name length for virtual nodes, as defined in the paper
+	// Must be a positive value
+	// TODO: Allow for the value of m to be changed after the FIB is created.
+	//       This would require deleting and recreating virtual nodes.
+	m int
+
+	// realTable is a map of names (hashed as uint64 values) to the FIB entry
+	// associated with that name.
+	realTable map[uint64]*baseFibStrategyEntry
+
+	// virtTable is a map of virtual names (hashed as uint64 values) to the
+	// virtualDetails struct associated with that virtual name.
+	virtTable map[uint64]*virtualDetails
+
+	// virtTableNames is a map of virtual names (hashed as uint64 values) to
+	// a set of all the real names associated with that virtual name. The
+	// inner map is being used as a set; the bool value type does not matter.
+	virtTableNames map[uint64](map[string]bool)
```

Comment:
(I'm not sure) If the only usage for this thing is to maintain `md`, actually you can use a data structure to maintain it directly, like put a priority queue into `virtualDetails` to maintain all max lengths.
Just some random thought, not required to change. FIB is not updated frequently and the updating performance does not matter that much.

Reply:
Well, it's OK. We don't need to modify.
But heap or BST(like red-black tree) can have duplicated values. It's not like C++ set or map.

Reply:
Oh you're right about the heap, don't know why I thought it couldn't handle duplicate values originally.